### PR TITLE
Check if `epoll_create1` is implemented by its error code

### PR DESCRIPTION
**Pull request by oblique**
If binary is statically linked to libc then `libc::dlsym` will always return `NULL`. This is the case when a target with musl is used.

To fix this we check if error code of `epoll_create1` is `ENOSYS`.

### CODE REVIEW

**Comment by oblique** [comment]
I marked it as draft because I think it will produce linkage error with glibc < 2.9. In that case I need to use `libc::syscall`. I will check this and let you know.

**Comment by oblique** [comment]
Ok ready

**Comment by stjepang** [comment]
Nice! So just to confirm - does this work now with glibc < 2.9?

@dignifiedquire This PR looks good to me, but just wanted to check with you because I believe you previously reported errors on old versions of Android that didn't have `epoll_create1`.

**Comment by oblique** [comment]
> Nice! So just to confirm - does this work now with glibc < 2.9?

I didn't find a system to test it but yes. `SYS_epoll_create1` is defined in `libc` crate as `c_long`. And kernel < 2.6.27 will return `ENOSYS` if system call number is invalid. So it does not depend at all in glibc.

**Comment by stjepang** [comment]
Maybe a silly question, but: is `SYS_epoll_create1` present even on unix systems that don't support it? My fear is that this would cause compilation errors due to `libc::SYS_epoll_create1` not being present, but I don't know if that should be a concern.

**Comment by oblique** [comment]
Yes it is present in all Linux systems as a number in libc crate.

**Comment by stjepang** [comment]
Ok great! In that case, I'm ready to merge this PR. :) I'm just gonna wait to see if @dignifiedquire is okay with merging too.

**Comment by dignifiedquire** [comment]
I am quite certain that trying to link to it will result in issues on the android systems I am trying to support. I won’t be close to a computer the next week, so can’t test it myself

**Comment by oblique** [comment]
@dignifiedquire No problem, we can wait. If indeed you have linkage error, paste it here.

File: src/epoll.rs [hunk 1]
```diff
@@ -30,48 +30,47 @@ impl Poller {
...
-                let epoll_create1 = std::mem::transmute::<
-                    *mut libc::c_void,
-                    unsafe extern "C" fn(libc::c_int) -> libc::c_int,
-                >(ptr);
-                match epoll_create1(CLOEXEC) {
-                    -1 => return Err(io::Error::last_os_error()),
-                    fd => fd,
+        //
+        // Use `epoll_create1` with `CLOEXEC`.
+        let epoll_fd = syscall!(syscall(libc::SYS_epoll_create1, CLOEXEC))
```
Comment (stjepang) [inline src/epoll.rs:35]: Is there any point in defining `const CLOEXEC` above if it's only used here? We could just put `libc::EPOLL_CLOEXEC` here, right?

File: src/epoll.rs [hunk 2]
```diff
@@ -151,41 +150,44 @@ impl Poller {
...
-        // Set interest in timerfd.
-        self.interest(
-            self.timer_fd,
-            Event {
-                key: crate::NOTIFY_KEY,
-                readable: true,
-                writable: false,
-            },
-        )?;
+            syscall!(syscall(libc::SYS_timerfd_settime, timer_fd, 0, &new_val, 0))?;
```
Comment (stjepang) [inline src/epoll.rs:166]: Let's be explicit with types to make sure we don't call the syscall with arguments of incorrect types:
```suggestion
            syscall!(syscall(libc::SYS_timerfd_settime, timer_fd, 0 as libc::c_int, &new_val as *const libc::itimerspec, 0 as *mut libc::itimerspec))?;
```

File: src/epoll.rs (hunk 1 above)
Comment (oblique) [reply src/epoll.rs]: correct

File: src/epoll.rs [hunk 3]
```diff
@@ -16,62 +16,57 @@ pub struct Poller {
...
-                // Use `epoll_create1` with `CLOEXEC`.
-                let epoll_create1 = std::mem::transmute::<
-                    *mut libc::c_void,
-                    unsafe extern "C" fn(libc::c_int) -> libc::c_int,
-                >(ptr);
-                match epoll_create1(CLOEXEC) {
-                    -1 => return Err(io::Error::last_os_error()),
-                    fd => fd,
+        //
+        // Use `epoll_create1` with `CLOEXEC`.
```
Comment (stjepang) [inline src/epoll.rs:27]:
```suggestion
        // Use `epoll_create1` with `EPOLLCLOEXEC`.
```

File: src/epoll.rs [hunk 4]
```diff
@@ -16,62 +16,57 @@ pub struct Poller {
...
+        let epoll_fd = syscall!(syscall(
+            libc::SYS_epoll_create1,
+            libc::EPOLL_CLOEXEC as libc::c_int
+        ))
+        .map(|fd| fd as libc::c_int)
+        .or_else(|e| {
+            match e.raw_os_error() {
+                Some(libc::ENOSYS) => {
+                    // If `epoll_create1` is not implemented, use `epoll_create`
+                    // and manually set `CLOEXEC`.
```
Comment (stjepang) [inline src/epoll.rs:37]:
```suggestion
                    // and manually set `FD_CLOEXEC`.
```

**Review by stjepang — APPROVED** [review]
We got a green light from @dignifiedquire so this is ready to merge. Thanks everyone!

**Pull request merged**